### PR TITLE
onMacroPatchesStateChanged func performance optimization

### PR DESCRIPTION
**Pull request by troublemaker52025**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by troublemaker52025** [comment]
@cocos-robot run test cases

**Comment by shaoqiangcai** [comment]
@cocos-robot run test cases

**Review by jk20012001 — APPROVED** [review]

File: native/cocos/scene/SubModel.cpp [hunk 1]
```diff
@@ -218,7 +220,16 @@ void SubModel::onPipelineStateChanged() {
 }
 
 void SubModel::onMacroPatchesStateChanged(const ccstd::vector<IMacroPatch> &patches) {
-    _patches = patches;
+    if (!patches.empty() && !patches.empty()) {
```
Comment (minggo) [inline native/cocos/scene/SubModel.cpp]: !patches.empty() && !patches.empty() -> patches.empty() && _patches.empty()

File: native/cocos/scene/SubModel.cpp [hunk 2]
```diff
@@ -218,7 +220,16 @@ void SubModel::onPipelineStateChanged() {
...
+        return;
+    }
+
+    ccstd::vector<IMacroPatch> tmp = patches;
+    std::sort(tmp.begin(), tmp.end(), IMacroPatch::compare);
+    if (std::equal(std::begin(tmp), std::end(tmp), std::begin(_patches))) {
```
Comment (minggo) [inline native/cocos/scene/SubModel.cpp]: I think this feature is c++17 depend, it is better to pass `std::end(_patches)` too. For example
```c++
std::equal(std::begin(tmp), std::end(tmp), std::begin(_patches), std::end(_patches))
```

File: native/cocos/scene/SubModel.cpp (hunk 1 above)
Comment (minggo) [reply native/cocos/scene/SubModel.cpp:223]: And i think it is better to change like this
```c++
if (_patches.empty()) _patches = patches;
else if (patches.empty()) _patches.clear();
else {
    // other logic;
}
```

Comment (minggo) [reply native/cocos/scene/SubModel.cpp]: And may be doesn't have to check it. If both of `patches` and `_patches` are empty, they can also use the same logic below. As empty equals empty.

File: native/cocos/scene/SubModel.cpp (hunk 2 above)
Comment (troublemaker52025) [reply native/cocos/scene/SubModel.cpp]: > ```c++
> , std::end(_patches)
> ```

sure

File: native/cocos/scene/SubModel.cpp [hunk 3]
```diff
@@ -218,7 +220,16 @@ void SubModel::onPipelineStateChanged() {
 }
 
 void SubModel::onMacroPatchesStateChanged(const ccstd::vector<IMacroPatch> &patches) {
-    _patches = patches;
+    if (patches.empty() && patches.empty()) {
```
Comment (minggo) [inline native/cocos/scene/SubModel.cpp]: patches.empty() -> _patches.empty()

Comment (minggo) [reply native/cocos/scene/SubModel.cpp]: Current code check `patches.empty()` twice.

**Review by minggo — APPROVED** [review]

**Pull request merged**